Add a contributing section to the sway book (#962)
FuelLabs/sway: docs/src/reference/contributing_to_sway.md

# Contributing To Sway

Thanks for your interest in contributing to Sway! This document outlines the process for installing and setting up the Sway toolchain for development, as well as some conventions on contributing to Sway.

If you run into any difficulties getting started, you can always ask questions on our [Discord](https://discord.gg/xfpK4Pe).

## Building and setting up a development workspace

> See the [introduction](../introduction/index.md) section for instructions on installing and setting up the Sway toolchain..

Reviewer: ```suggestion
See the [introduction](../introduction/index.md) section for instructions on installing and setting up the Sway toolchain.
```

## Getting the repository

1. Visit the [Sway](https://github.com/FuelLabs/sway) repo and fork the project.
2. Then clone your forked copy to your local machine and get to work.

```sh
git clone https://github.com/FuelLabs/sway
cd sway
```

## Building and testing

The following steps will run the sway test suite and ensure that everything is set up correctly.

> First, open a new terminal and start `fuel-core` with:

Reviewer: Maybe prior to this we should have a section on installing `fuel-core` along with a brief sentence explaining why it's required to run the tests? Or perhaps provide a link if we already have a guide on `fuel-core` setup elsewhere?

Reviewer 2: The Sway Book should already talk about installing `fuel-core`, so a link should be sufficient.

Author: The section above, `Building and setting up a development workspace` already has a link to the sway book's [introduction](https://fuellabs.github.io/sway/latest/introduction/index.html) which _should_ be sufficient?

- Reporting bugs
- Adding documentation to the Sway book
- Adding new features or bugfixes for which there is already an open issue
- Making feature requests

Check out our [Help Wanted](https://github.com/FuelLabs/sway/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22), [Sway Book](https://github.com/FuelLabs/sway/issues?q=is%3Aopen+is%3Aissue+label%3A%22The+Sway+Book%22) or [Good First Issue](https://github.com/FuelLabs/sway/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22) issues to find a suitable task.

If you are planning something big, for example, related to multiple components or changes current behaviors, make sure to open an issue to discuss with us before starting on the implementation.

## Contribution flow

This is a rough outline of what a contributor's workflow looks like:

- Make sure what you want to contribute is already tracked as an issue.
  - We may discuss the problem and solution in the issue.
- Create a Git branch from where you want to base your work. This is usually master.
- Write code, add test cases, and commit your work.
- Run tests and make sure all tests pass.

> - If the PR contains any breaking changes, add the breaking label to your PR.

Reviewer: Let's remember to add a "breaking" label for the sway repo after landing this.

# Contributing To Sway

> Thanks for your interest in contributing to Sway! This document outlines the process for installing and setting up the Sway toolchain, as well as some conventions on contributing to Sway.

Reviewer 2: ```suggestion
Thanks for your interest in contributing to Sway! This document outlines the process for installing and setting up the Sway toolchain for development, as well as some conventions on contributing to Sway.
```

If you run into any difficulties getting started, you can always ask questions on our [Discord](https://discord.gg/xfpK4Pe).

## Building and setting up a development workspace

> See the Sway book's [Introduction](../introduction/index.md) for instructions on installing and setting up the Sway toolchain..

Reviewer 2: You're already in The Sway Book, so no need for saying "see the Sway book's..."

## Getting the repository

> ```

Reviewer 2: And elsewhere
````suggestion
```sh
````

> git clone https://github.com/FuelLabs/sway

Author: Should we be directing contributors to fork the sway repo first?

Reviewer 2: Probably, yes

cd sway
```

## Building and testing

The following steps will run the sway test suite and ensure that everything is set up correctly.

> First, open a new terminal and start fuel-core with:

Reviewer 2: ```suggestion
First, open a new terminal and start `fuel-core` with:
```

```
fuel-core
```

Then, open a second terminal and run:

```

> cd sway/test
> cargo run

Reviewer 2: Shorter would be
```suggestion
cargo run --bin test
```

```

After the test suite runs, you should see:

> > ---
> >
> > Tests passed.\
> > _n_ tests run (0 skipped)

Reviewer 2: ````suggestion
```console
Tests passed.
_n_ tests run (0 skipped)
```
````

Congratulations! You've now got everything setup and are ready to start making contributions.

## Finding something to work on

There are many ways in which you may contribute to the Sway project, some of which involve coding knowledge and some which do not. A few examples include:

- Reporting bugs
- Adding documentation to the Sway book
- Adding new features or bugfixes for which there is already an open issue

> - making Feature requests

Reviewer 2: ```suggestion
- Making feature requests
```

Check out our [Help Wanted](https://github.com/FuelLabs/sway/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22), [Sway Book](https://github.com/FuelLabs/sway/issues?q=is%3Aopen+is%3Aissue+label%3A%22The+Sway+Book%22) or [Good First Issue](https://github.com/FuelLabs/sway/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22) issues to find a suitable task.

> If you are planning something big, for example, related to multiple components or changes current behaviors, make sure to open an issue to discuss with us before going on.

Reviewer 2: ```suggestion
If you are planning something big, for example, related to multiple components or changes current behaviors, make sure to open an issue to discuss with us before starting on the implementation.
```

## Contribution flow

This is a rough outline of what a contributor's workflow looks like:

- Make sure what you want to contribute is already tracked as an issue.
  - We may discuss the problem and solution in the issue.
- Create a Git branch from where you want to base your work. This is usually master.
- Write code, add test cases, and commit your work.
- Run tests and make sure all tests pass.
- If the PR contains any breaking changes, add the breaking label to your PR.
- Push your changes to a branch in your fork of the repository and submit a pull request.
  - Make sure mention the issue, which is created at step 1, in the commit message.
- Your PR will be reviewed and some changes may be requested.
  - Once you've made changes, your PR must be re-reviewed and approved.
  - If the PR becomes out of date, you can use GitHub's 'update branch' button.
  - If there are conflicts, you can merge and resolve them locally. Then push to your PR branch.
    Any changes to the branch will require a re-review.
- Our CI system (Github Actions) automatically tests all authorized pull requests.
- Use Github to merge the PR once approved.

Thanks for your contributions!

### Linking issues

> Pull Requests should be linked to at least one issue in the same repo.

Reviewer 2: ```suggestion
Pull requests should be linked to at least one issue in the same repo.
```
